Add busy favicon when kernels are running

saulshanabrook opened the pull request:
Fixes #3957 by changing the favicon when kernels are running. 


As implemented, this uses the session manager to update, but this is a bit slow. It isn't refreshed as soon as the kernel status changes, but it does seem to work.

It would might be quicker if we subscribe to the `ISession.statusChanged` or the `IKernel.statusChanged` signals instead of the `SessionManager.runningChanged` signal.

However, I am not sure where the best place to react to an updated global list of `ISession`'s or `IKernel`'s is.

ian-r-rose commented on packages/application-extension/src/index.tsx:
```diff
@@ -151,6 +151,30 @@ const main: JupyterLabPlugin<void> = {
+        console.log({isBusy});
+        if (isBusy && !interval) {
+          interval = window.setInterval(() => {
+              favicon.href = `/static/base/images/${icons[i++ % 3]}`;
```
Are you animating the favicon here? What does the classic notebook use the other busy indicators for? I only see references to `favicon-busy-1.ico`.

saulshanabrook replied:
Ah, they used to use the animated favicon, but then they removed it: https://github.com/jupyter/notebook/pull/2675/files For some reason I was found the commit adding the animated state and didn't see that they undid it.

I will change it not to animate.

saulshanabrook commented:
@ian-r-rose this is ready for another review whenever you have time

ellisonbg commented:
Can you post a screenshot?

ian-r-rose commented on packages/application/src/shell.ts:
```diff
@@ -657,6 +664,25 @@ class ApplicationShell extends Widget {
     this._currentChanged.emit(args);
     this._onLayoutModified();
+
+    this._onCurrentSession(args.newValue ? (args.newValue as any).session : undefined);
```
Do we have a more type-safe way of getting at whether there is an `IClientSession` associated with a widget? This works for notebooks and consoles, but more-or-less by accident, in that they both have the `session` available by that name on the top-level widget. If another widget comes along with a different API, but still uses an `IClientSession` this will not work.

I can think of how we would do it for documents: we could use the `DocumentManager.contextForWidget` function to get the context, which has a `session` object on it. I don't currently know how to do it for a more general widget...

ian-r-rose replied:
Ah, sorry, I missed that you addressed this in the above comment. Yeah, I am not sure about this just now (though I think the docmanager solution would work for documents).

saulshanabrook replied:
Is there some way for a widget to register a `isBusy` for itself? So not base it on the session object at ll but make it more generic? Like there is the `setDirty` method on the application that widgets can use to mark if they are unsaved. The simplest way I can think of is storing a [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) in the application mapping widgets to `isBusy` signals. Any widget could add itself and a signal to this map when it is created.

ian-r-rose replied:
I like this idea! I think it could follow a very similar logic to the `setDirty` setup (which doesn't use a `WeakMap`, but instead a `DisposableDelegate` with an internal count of the dirty widgets.

ian-r-rose commented on packages/application/src/shell.ts:
```diff
@@ -657,6 +664,25 @@ class ApplicationShell extends Widget {
+  /**
+   * Handle a change to the session of the current widget.
+   */
+  private _onCurrentSession(session?: IClientSession) {
```
Can we change these to `_onCurrentSessionChanged` and `_onCurrentStatusChanged`, or similar? I found the names here to be kind of confusing.

saulshanabrook replied:
Yep, sounds good, will do.

saulshanabrook commented:
@ellisonbg Here is a quick gif showing how it changes based on selected widget:

![](http://g.recordit.co/PWFGjBbfmh.gif)

saulshanabrook commented:
I am thinking of just implementing this as a global check if any kernel is busy, like the `isDirty` flag, since that's a bit simpler. Please weigh in if you feel strongly that it should show the specific kernel of the open window.

jasongrout commented:
I think it is more understandable and less confusing if it is just a global check.

saulshanabrook commented:
I have updated it to be a global check.

It handles updates from notebooks:

![](http://recordit.co/HOzhOkCUN9.gif)

And updates from consoles:

![](http://recordit.co/tU5FbXqdFe.gif)

It also takes into account the status of kernels after a reload, for kernels that aren't on screen:

![](http://recordit.co/ECnbBjkd5W.gif)

I think there is currently a bug when a window is reloaded with a kernel currently executing, where it is seen as not busy on reload. So this affects the favicon as well:

![](http://recordit.co/ex4T938bKm.gif)

It does this by combining the `/api/kernels` API call with the frontend kernel status changes. This let's us refresh from server status every couple of seconds and live update with the local changes.

ian-r-rose commented:
Why did you not do the `isBusy` approach that we talked about earlier? We are trying to be as conservative as possible with `@jupyterlab/services`, not changing the API surface area unless we really need to. Another thing that is nice about the `isBusy` approach is that it is usable by other extensions (even ones that have concepts of "busy" that are not necessarily tied to a given kernel).

saulshanabrook commented:
This makes use of the backend response, so if we have a kernel running that isn't displayed it will still register as busy.

I supposed that could be added with the previous approach though. I thought this was cleaner, so then we don't have to have special cases for notebooks and consoles.

saulshanabrook commented:
All this back and forth makes me realize I should have properly scoped out this feature before implementing it in different ways!

ian-r-rose commented:
I think that showing a busy signal for a running kernel that is not tied to a given activity might be kind of confusing.

ian-r-rose commented:
Yes, all of these decisions that the classic notebook has made suddenly become a lot subtler when you can have many activities on the same page! What do you think @jasongrout?

jasongrout commented:
> All this back and forth makes me realize I should have properly scoped out this feature before implementing it in different ways!

Part of the back-and-forth and part of the implementation experimentation is figuring out the scope, so I'm not sure the time was for nought.

saulshanabrook commented:
> Part of the back-and-forth and part of the implementation experimentation is figuring out the scope, so I'm not sure the time was for nought.

And also helpful for myself to learn about the codebase

jasongrout commented:
> Yes, all of these decisions that the classic notebook has made suddenly become a lot subtler when you can have many activities on the same page! What do you think @jasongrout?

How about we answer some usecase-based questions, maybe on the original thread where we have outside engagement.

saulshanabrook commented:
@jasongrout that would be great. Thank you.

ellisonbg commented:
In our dev call today:

* We like the idea of having an application wide a `isBusy` attribute handled similarly to `isDirty`
* We like the idea of a global notion of being busy.

ellisonbg commented:
And the notebook/console extensions would use this...

saulshanabrook commented:
@jzf2101 Could you try again? I just rebased and implemented the changes we discussed last week, i.e. changing to using a `setBusy` function on the application. I ended up calling that in the clientsession, instead of in the console panel and notebook panel, since then we just have to write the logic once.

jzf2101 commented:
@tgeorgeux do you want to look at it before we merge?

ian-r-rose commented:
I'd still like to look over the most recent version.

saulshanabrook commented:
@ian-r-rose Thanks for the review. I am inclined to let his merge first, to make it easiest for him to get that in ASAP. There isn't a ton here, so rebasing shouldn't be that bad.

saulshanabrook commented on packages/apputils/src/clientsession.tsx:
```diff
@@ -739,6 +740,23 @@ class ClientSession implements IClientSession {
+      } else {
+        if (this._busyDisposable) {
+          this._busyDisposable.dispose();
+          delete this._busyDisposable;
```
For example, when the session first starts there is no `_busyDisposable`, so if we get some status, we shouldn't try to dispose of it. We should only dispose of it when it exists.

ian-r-rose replied:
It should be safe to dispose of it twice, I think. But if we are worried about that, can we explicitly set it to `null` instead?

ian-r-rose replied:
Ah, I see what you mean. Yeah, I think it would be more consistent with the rest of the codebase to set the class member to `null` (we currently are only using the `delete` operator on plain `Object`s, rather than ES6 classes).

saulshanabrook replied:
Sure, just pushed.

jasongrout commented:
I would love to merge #4453 first, if possible.

ian-r-rose commented:
Looks like no merge conflicts, so full steam ahead!
